py/server/deephaven/table_reader.py

```diff
@@ -0,0 +1,91 @@
+#
+# Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+#
+"""This module supports reading the data in a Deephaven table in a chunked manner."""
+
+from typing import Union, Sequence, Generator, Dict, Optional
+
+import jpy
+import numpy as np
+
+from deephaven.column import Column
+from deephaven.jcompat import to_sequence
+from deephaven.numpy import _column_to_numpy_array
+from deephaven.table import Table
+
+_JTableUpdateDataReader = jpy.get_type("io.deephaven.integrations.python.PythonTableDataReader")
+
+
+def _col_defs(table: Table, cols: Union[str, Sequence[str]]) -> Sequence[Column]:
+    if not cols:
+        col_defs = table.columns
+    else:
+        cols = to_sequence(cols)
+        col_defs = [col for col in table.columns if col.name in cols]
+
+    return col_defs
+
+def table_reader(table: Table, cols: Optional[Union[str, Sequence[str]]] = None, *, chunk_size: Optional[int] = None) \
+        -> Generator[Dict[str, np.ndarray], None, None]:
+    """ A generator that reads the table in chunks of rows. It yields a dictionary of column names to numpy arrays.
+
+    Note, If the table is refreshing, the table reader is not guaranteed to see a consistent view of the data. To ensure
```

Comment:
This makes me super nervous given typical user behavior.  It seems like a big footgun.  Is it possible to log a console warning if the code is being executed (1) not under a lock and (2) not on a static table?

Should there be an annotation to grab the lock if it is not already held?

Are there any use cases where (1) there is no lock or (2) the table is not static, or are all such cases bugs?